### PR TITLE
Xspuserprofilesection

### PR DESCRIPTION
**Pull request by camiloborges**
Adding support to manage user profile sections.

@

### CODE REVIEW

File: Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSection/MSFT_xSPUserProfileSection.psm1 [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        $userProfileSubType = $userProfileSubTypeManager.GetProfileSubtype("UserProfile")
+        
+        $userProfileProperty = $userProfileSubType.Properties.GetSectionByName($params.Name) 
+        if($userProfileProperty -eq $null){
+            return $null 
+        }
+        
+        
+        return @{
+            Name = $userProfileProperty.Name
```
Comment (TravisEz13) [inline Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSection/MSFT_xSPUserProfileSection.psm1]: Should probably also return ensure so I don't have to infer if the section is present or absent.

File: Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+                $Global:xSPUPGetSectionByNameCalled = $false
+                $Global:xSPUPCoreRemoveSectionByNameCalled=$false
+                Set-TargetResource @testParams 
+                $Global:xSPUPGetProfileSubtypeCalled | Should be $true
+                $Global:xSPUPGetSectionByNameCalled | Should be $true
+                $Global:xSPUPCoreRemoveSectionByNameCalled | Should be $true
+            }           
+        }
+    }    
+}
```
Comment (TravisEz13) [inline Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1]: Files must end with a new line.  See [Style Guidelines, General Rule, #5](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md#general-rules)

File: Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        
+        #$typeProperties.Add($typeProperty)
+       $subTypePropertyUpdate = @{
+                            Name = "PersonalInformation"
+                            DisplayName = "Personal InformationUpdate"
+                            DisplayOrder =5000
+                        }| Add-Member ScriptMethod Commit {
+                            $Global:xSPUPPropertyCommitCalled = $true
+                        } -PassThru 
+                        #>
```
Comment (BrianFarnhill) [inline Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1]: Can you pull out this commented stuff if its not needed?

**Comment by BrianFarnhill** [comment]
@camiloborges  verall script looks fine - just need to review the comments from both myself and @TravisEz13 then do a fresh pull from dev so we can merge it mate

File: Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1 (hunk 3 above)
Comment (camiloborges) [inline Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1]: Sure!
Might be able to do it soon. 
Camilo Borges
Sent via iPhone while out and about

**Comment by camiloborges** [comment]
hey. bit of issue with merge again, but I don't think I've screwed up the PR this time around.

please don't merge just yet, I'd like to run few more integration tests., will do tomorrow

**Comment by BrianFarnhill** [comment]
@camiloborges You happy with this one yet mate? Just wanted to be sure before I review again

**Comment by camiloborges** [comment]
hey. I finally got a chance to check this out. Found a bug, whereby DSC seems to work fine when Service App isn't available. 
It is now going to throw an exception if service isn't found. 
(probably same issue exists in user profile property)

**Comment by camiloborges** [comment]
yo, sorry. Ended up refactoring and simplifying the module, so need to make changes to the unit tests. hopefully will get it done tomorrow

**Comment by camiloborges** [comment]
@BrianFarnhill go nuts. I've managed to simplify the code a fair bit without changing the functionality. Great result.

File: Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSection/MSFT_xSPUserProfileSection.psm1 [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        $upsa = Get-SPServiceApplication -Name $params.UserProfileService -ErrorAction SilentlyContinue 
+        if ($null -eq $upsa) { 
+            return $null 
+        }
+        $caURL = (Get-SpWebApplication  -IncludeCentralAdministration | ?{$_.IsAdministrationWebApplication -eq $true }).Url
+        $context = Get-SPServiceContext -Site $caURL 
+        $userProfileConfigManager  = new-object Microsoft.Office.Server.UserProfiles.UserProfileConfigManager($context)
+        
+        #$userProfileSubTypeManager = Get-xSharePointUserProfileSubTypeManager $context
+        #$userProfileSubType = $userProfileSubTypeManager.GetProfileSubtype("UserProfile")
```
Comment (BrianFarnhill) [inline Modules/xSharePoint/DSCResources/MSFT_xSPUserProfileSection/MSFT_xSPUserProfileSection.psm1]: Need to remove the commented out stuff here mate

File: Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        $coreProperties = @{ProfileInformation = $coreProperty}
+        $userProfileSubTypePropertiesProperty = @{"ProfileInformation" = $subTypeProperty } | Add-Member ScriptMethod Create {
+                            $Global:xSPUPSubTypeCreateCalled = $true
+                        } -PassThru | Add-Member ScriptMethod Add {
+                            $Global:xSPUPSubTypeAddCalled = $true
+                        } -PassThru -Force <#| Add-Member ScriptMethod GetSectionByName {
+                            $Global:xSPUPGetSectionByNameCalled  = $true
+                            return $subTypeProperty
+                        } -PassThru
+                        #>
```
Comment (BrianFarnhill) [inline Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1]: More comments here - I take it this isn't actually needed?

Comment (camiloborges) [inline Tests/xSharePoint/xSharePoint.xSPUserProfileSection.Tests.ps1]: Yeah, they should all be gone. I though I had removed them. Will check it out when I get home. 


Camilo Borges
Sent via iPhone while out and about

**Comment by camiloborges** [comment]
hey Brian, we should be good to include this in the latest release as well. There is a small UPS Property bug fixed here too

**Pull request merged**